audit: zap3 Fixed issues

Doublo54 opened the pull request:
# ApeSwap-hardhat-zap Paladin audit review

### Issue 4: removeLiquidityGamma is flawed
- [x] Resolved: Transferring in shares first before withdrawing
GammaHypervisor requires the removal to be called `from` owner. Why would they do that?
https://polygonscan.com/address/0x02203f2351e7ac6ab5051205172d3f772db7d814#code
- [ ] Tested

### Issue 23: 
- [x] Resolved
The way the liquidity add is calculated on Arrakis differs depending on if there is already liquidity or not: https://github.com/ArrakisFinance/vault-v1-core/blob/7a06f6ffa744eee39b86db9eabb4ea67d3c98aec/contracts/ArrakisVaultV1.sol#L678
- [ ] Test: `estimateSwapReturns` for Arrakis

### Issues to revisit 
#### Issue 10: The amount parameter can differ from msg.value
- [x] Resolved
Added check for `msg.value >= amount` in case there is a use case where wrap is less than msg.value (don’t know a use case yet)

#### Issue 17: `msg.value` is unchecked 
- [x] Resolved: Added `noNativeSurplus` modifier to `multicall` to prevent excess BNB from being left in contract. 

### Issues 6, 11, 12, 13, 14, 15, 16, 18, 22, 24, 25, 27, 28
- [x] Resolved

## Issues with comments (Resolved)
### Issue 1: Token recipients can be address(this) 
Fixed for all except ZapLiquidity because LPs can be zapped to pools, vaults etc

### Issue 3: Lack of validation for constructor arguments
Fixed except for banana treasury because zero address is allowed if not bnb chain. Validation for bnb chain is in ApeSwapZapPool

### Issue 8: Typographical issues
IGammaUniProxy and IGammaHypervisor are used. Renamed UniProxy and Hypervisor interfaces to match file name.
For AddLiquidityV2Params, lpRouter now defined as IV2LiquidityRouter02
Fixed IArrakisRouter lpRouter 


### Issue 19: GNANA logic is flawed
Fixed by adding check if inputToken == GNANA then transferIn Banana and else just the token like normal.
- [x] Resolved

## Discussion (Not resolved)
### Issue 2: Unnecessary payable functions
Payable necessary for multicall

### Issue 5: Tokens with a fee on transfer will lose value
Aware and noted. Probably not planning to use such tokens here?

### Issue 7: Several functions accept native value when they should not
Payable necessary for multicall

### Issue 9: The unwrapNative function is flawed
Usage only meant in combination with other functions in multicall. Example: swap busd->wbnb, unwrap and withdraw. Maybe we just shouldn’t allow unwrap and keep in contract. First it could be used for zapping to lending but with changes not anymore.

### Issue 20: estimateSwapReturns might run out of gas
No code changes. “Consider monitoring this issue and prepare a custom RPC node that can deal with a custom gas limit.”

### Issue 21: Discrepancy between router and factory
We are aware and don’t think it’s a big deal. Just depends on the liquidity type that we need to pass.

### Issue 26: getArrakisPool might run out of gas
“Due to the logic in the Arrakis contracts, there is no easy way to solve this issue“. Noted but not sure how much we can do here (and not using Arrakis (yet))

### Issue 29: getLPAddRatio for Gamma can be different from result
We have to choose something here. Right in the middle makes the most sense to me. Least chance of reverting because of slippage.

### Issue 30: Missing warning for payable modifier
Not added acknowledgment into the code yet. It’s only vulnerable to using msg.value in code which we have in 2 places. wrapNative and ApeSwapZapLending. In both places, you can steal eth if it’s in the contract. But zap is not designed to actually hold any eth and it’s recommended to only do multicalls where ALL funds are used and transferred back. So should not be a (big) issue? But let’s discuss.

Doublo54 committed:
audit: Fixed issues

Doublo54 committed:
remove old contracts from contracts folder so they don't compile

Doublo54 committed:
audit: Fixed estimateSwapReturns?

DeFiFoFum commented on contracts/lens/ZapAnalyzer.sol:
```diff
@@ -250,23 +252,20 @@ contract ZapAnalyzer is IZapAnalyzer {
             SwapPath memory path = fullPath[i];
             if (path.swapType == SwapType.V2) {
+                //divide by 1/6 to decrease slippage impact. Noted that it only works for tokens with 6 or more decimals
+                uint256 amount = 10**IERC20Metadata(path.path[0]).decimals() / 1e6;
```
Why is it important to decrease slippage impact when checking weightedPrice? We have done bonds with a token with a payout token of 3 decimals before which wouldn't work for this.

Doublo54 replied:
Lets take btc-quick as example. Btc at $30000 and quick at $0.05. lets say there's $60k worth in pool. $30k btc (1) and $30k quick (600.000). First it tried estimating swapping numbers with 1 full btc with high slippage. So it estimates like 200.000 quick output instead of the 600.000.
So maybe you add just 0.001 btc and thus within zap 2000 quick to the lp.
With estimation of 1e12 btc the slippage is less and you add 0.001 btc with 5900 quick which is way closer the actual ratio. Just random guessing example numbers but hope you get the point. And for gamma it just failed before because it needs be inside a specific small ratio

DeFiFoFum replied:
Ok, this is helpful thanks. 

I added something more dynamic like this:
```
                uint256 precision = 1e6; // Use a precision of 1e6 for tokens with more than 12 decimals
                if (fullTokenAmount <= 1e12) {
                    precision = 1e4; // Use a precision of 1e4 for tokens with more than 6 decimals
                } else if (fullTokenAmount <= 1e6) {
                    precision = 1e2; // Use a precision of 1e2 for tokens with less than or equal to 6 decimals
                }
```

DeFiFoFum commented on contracts/utils/TokenHelper.sol:
```diff
@@ -32,14 +32,14 @@ library TokenHelper {
             decimals = dec;
         } catch {
-            decimals = 18;
+            revert("Token decimals can't be retrieved");
```
Issue #27 Recommendation: "Consider if this becomes an issue, and if so, consider simply reverting in that case." 

The point of using a try catch was to not mess up the flow when trying to pull decimals from a token. Where would this become an issue?

Doublo54 replied:
A token which hardcodes the decimals without having a decimals() retrieving function and is not 18 decimals. But yeh what are the odds of that happening. And we can make sure to not use weird tokens on the fe. I'm good changing it back

DeFiFoFum commented on contracts/extensions/lending/ApeSwapZapLending.sol:
```diff
@@ -11,36 +11,43 @@ abstract contract ApeSwapZapLending is TransferHelper {
         if (address(underlyingToken) == LENDING_NATIVE_UNDERLYING) {
-            uint256 depositAmount = inputAmount == Constants.CONTRACT_BALANCE ? address(this).balance : inputAmount;
-            market.mint{value: depositAmount}();
+            params.market.mint{value: msg.value}();
```
I'm confused about this update because doesn't this prevent a user from zapping into lending? 

1. They use an input token other than a Native asset
2. This means that `msg.value == 0`
3. Trying to zap into the Native market would revert because there is no msg.value?

Doublo54 replied:
"It does limit it a bit. For example technically first you could multicall (swap +) unwrap and zapLendingMarket. This is not possible anymore."

Yeh true, mentioned that. Don't exactly remember but there was a Safety issue here if we didn't?

DeFiFoFum replied:
We can't normally trust `msg.value` in a Multicall context, but I added checks for both in and out of Multicall context in the `MulticallGuard`

DeFiFoFum commented on contracts/WrapNative.sol:
```diff
@@ -5,22 +5,29 @@ import "./interfaces/IWETH.sol";
 
     function wrapNative(uint256 amount, address recipient) external payable {
+        require(recipient != address(0), "ApeSwapZap: recipient can't be address()");
+        require(msg.value >= amount, "ApeSwapZap: msg.value must be higher than or equal to amount to wrap");
```
- [ ] Return any extra `msg.value` to the user so it doesn't get locked in the contract

Doublo54 replied:
If we add that we might as well make it msg.value == amount. 
The point of >= was to keep it more flexible in the multicall so you can do stuff like send 1 eth, swap 0.5 to some lp and the other 0.5 to lending (which I guess the other change doesn't make possible now).
So either just make it == or keep this flexibility with risk of leaving funds in there

DeFiFoFum replied:
ahh I see, I like the idea of using `>=` in that case, definitely want the flexibility. 

Maybe we can add like `beforeZap` and `afterZap` hooks which allow us to double check that no BNB was left in the contract 🤔

Doublo54 replied:
Ohh I like that. Does just after the multicall make sense? 🤔

DeFiFoFum replied:
I added this into the `MulticallGuard` contract which checks the balance before and after.

DeFiFoFum commented on contracts/extensions/pools/ApeSwapZapPools.sol:
```diff
@@ -49,17 +49,25 @@ abstract contract ApeSwapZapPools is TransferHelper {
-        else if (params.recipient == Constants.ADDRESS_THIS) params.recipient = address(this);
 
         if (inputToken == GNANA) {
+            params.inputAmount = _transferIn(BANANA, params.inputAmount);
```
Need to double check this as wouldn't this mean that we couldn't deposit tokens which are already in the contract.

Doublo54 replied:
When inputAmount is 0 (and inputtoken is gnana) you get all banana from the contract. Or is that not what you mean?

DeFiFoFum replied:
Ahh ok nice, I forget about the case where it can use the contract balance. 👌

DeFiFoFum commented on contracts/extensions/pools/ApeSwapZapPools.sol:
```diff
@@ -49,17 +49,25 @@ abstract contract ApeSwapZapPools is TransferHelper {
             GNANA_TREASURY.buy(params.inputAmount);
             params.inputAmount = inputToken.balanceOf(address(this)) - beforeAmount;
+        } else {
+            params.inputAmount = _transferIn(inputToken, params.inputAmount);
```
(See comment above)

Doublo54 replied:
Same? ;)

DeFiFoFum committed:
audit: Resolve issues 4 and 23
- removeLiquidityGamma: Resolve by transferring in tokens first
- Use `estimateSwapReturns` to estimate the amounts for Arrakis LP add
- Refactor check for ZapAnalyzer in constructor of Zap
- Refactor precision calculation in _getWeightedPrice

DeFiFoFum committed:
refactor: TokenHelper, fallback to 18 decimals

DeFiFoFum committed:
feat: Add noNativeSurplus modifier to ensure there is no extra native balance left in contract
- This helps ensure that users don't send excess Native tokens to the contract they didn't intend on which could be captured by other users
- Refactored ApeSwapZapLending to allow for input amount over msg.value

DeFiFoFum committed:
docs: Provide additional dev documentation.
- Organize functions to match similar flow for readability

DeFiFoFum committed:
docs: Provide additional dev documentation.
- Organize functions to match similar flow for readability

DeFiFoFum committed:
feat: Add Multicall Guard
- Protects against reentrancy attacks during multicall
- Protects against reentrancy attacks out of multicall
- Provides helper functions for validating checks depending on the state of the multicall

DeFiFoFum committed:
refactor: Integrate MulticallGuard into Zap Features

DeFiFoFum committed:
refactor: Add additional protections
- Add `receive` function to protect against unwanted native deposits
- Rename SwapParams to be more clear
- Reset approvals after swap() call

DeFiFoFum committed:
chore: Lint fix

DeFiFoFum committed:
refactor: Add reentrant check for guardMulticall
- Fix typography

DeFiFoFum committed:
feat: Add DeployManager and deploy.config for Zap deployments
- Update hardhat-etherscan to version 3

DeFiFoFum committed:
security: Patch

DeFiFoFum committed:
Merge branch 'feat/zap3-audit' into feat/zap3-audit-multicall-validation

DeFiFoFum committed:
Merge pull request #9 from ApeSwapFinance/feat/zap3-audit-multicall-validation

feat: Add Multicall Guard
